
## Data preparation

Ð¡onvert data to csv format to read it easily with R and remove unuseful columns from it

```{r}
#data <- read.csv("../data/IAD30284_384WellPlateDataSheet.csv", header = TRUE, skip = 14)
data <- read.csv("../data/panel2/IAD39777_51_384WellPlateDataSheet.csv", header = TRUE, skip = 9)
data <- data[data$Amplicon_ID != "Blank", ]
data$X384Well_Row <- NULL
data$X384Well_Col <- NULL
data$Gene_Symbol <- NULL
data$Genome_Version <- NULL
data$LineItem_Type <- NULL
data <-data[complete.cases(data), ]
```

Get amplicon sequence that lies between primers
```{r}
library(jsonlite)
library(httr)
n <- nrow(data)
chromosomes <- data$Chr
beginings <- data$Insert_Start
endings <- data$Insert_Stop
ampliconInnerPart <- rep("", n)
ampliconInnerPart10 <- rep("", n)
unvalid <- NULL
for (i in 1:n)
  if (endings[i]-beginings[i]<=4000) {
    call <- paste(c("http://178.79.134.178/refservice/references/GRCh37.p13/sequence?contig=", as.character(chromosomes[i]),
                           "&start=",as.character(beginings[i]),"&end=",as.character(endings[i])), collapse = "")
    jsonData <- fromJSON(call)
    ampliconInnerPart[i] <- jsonData$sequence
    ampliconInnerPart10[i] <- substr(jsonData$sequence, 10, 10)
  }else{
    unvalid <- c(unvalid, i)
  }
print(unvalid)
data <- data[-unvalid, ]
ampliconInnerPart <- ampliconInnerPart[-unvalid]
ampliconInnerPart10 <- ampliconInnerPart10[-unvalid]
head(ampliconInnerPart)
data$Amplicon_Inner_Part <- ampliconInnerPart
data$Amplicon_Inner_Part10 <- ampliconInnerPart10
```

Get dangling ends
```{r}
n <- nrow(data)
danglingEndStart <- rep("", n)
danglingEndStop <- rep("", n)
for (i in 1:n){
  call <- paste(c("http://178.79.134.178/refservice/references/GRCh37.p13/sequence?contig=", as.character(chromosomes[i]),
                         "&start=",as.character(beginings[i]-1),"&end=",as.character(beginings[i]-1)), collapse = "")
  jsonData <- fromJSON(call)
  danglingEndStart[i] <- jsonData$sequence
  call <- paste(c("http://178.79.134.178/refservice/references/GRCh37.p13/sequence?contig=", as.character(chromosomes[i]),
                        "&start=",as.character(endings[i]+1),"&end=",as.character(endings[i]+1)), collapse = "")
  jsonData <- fromJSON(call)
  danglingEndStop[i] <- jsonData$sequence
}
data$Start_Dangling_End <- danglingEndStart
data$Stop_Dangling_End <- danglingEndStop
```

```{r, echo =FALSE}

readResultFile <- function(name, addName = FALSE, delim = "\t"){
  filePath <- paste(c("../data/panel2/", name), collapse = "")
  result <- read.table(filePath, sep = delim, header = TRUE)
  result$Gene <- NULL
  result$sam <- NULL
  result$X <- NULL
  sums <- colSums(result[,-1])
  proportion_result <- sweep(result[,-1], 2, sums, '/' )
  #proportion_result <- sapply(proportion_result, function(x) log(x))
  proportion_result$Means <- rowMeans(proportion_result)
  proportion_result$Target <- result$Target
  y <- data.frame(Target = character(nrow(result)))
  y$Target <- result$Target
  ylog <- data.frame(Target = character(nrow(result)))
  ylog$Target <- result$Target
  if (addName){
    logresultStr <- paste(c(name,"_logresult"), collapse = "")
    logMeansStr <- paste(c(name,"_logMeans"), collapse = "")
    resultStr <- paste(c(name,"_result"), collapse = "")
    MeansStr <- paste(c(name,"_Means"), collapse = "")
  }else{
    logresultStr <- "logresult"
    logMeansStr <- "logMeans"
    resultStr <- "result"
    MeansStr <- "Means"
  }
  #ylog[resultStr] <- log(proportion_result[,1])
  ylog[MeansStr] <- log(proportion_result$Means)
  #y[resultStr] <- proportion_result[,1]
  y[MeansStr] <- proportion_result$Means
  ylog[ylog[MeansStr] == -Inf,][MeansStr] <- -10 # min(y[ylog$Means != -Inf,]$logMeans) - 20
  #ylog[ylog[resultStr] == -Inf,][resultStr] <- -10 # min(y[ylog$result != -Inf,]$logresult) - 20
  rownames(y) <- y$Target
  rownames(ylog) <- ylog$Target
  proportion <- proportion_result
  proportion$Means <- NULL
  proportion$Target <- NULL
  rownames(proportion) <- proportion$Target
  return(list(y = y, ylog = ylog, proportions = proportion))
}
```

Load results and get a proportion for each
```{r}
res <- readResultFile("Run_14.xls",TRUE)
totalTable <- data.frame(Target = character(nrow(res$proportions)))
totalTableOfMeans <- data.frame(Target = character(nrow(res$proportions)))
totalTableOflogMeans <- data.frame(Target = character(nrow(res$proportions)))
namesOfColumns <- names(res$proportions)
totalTable[namesOfColumns] <- res$proportions[namesOfColumns]
namesOfColumns <- names(res$y)
totalTableOfMeans[namesOfColumns] <- res$y
totalTableOflogMeans[namesOfColumns] <- res$ylog
y <- res$y
ylog <- res$ylog
```

Load results and get a proportion for each
```{r}
#resultFileNames <- c("SN1-17_Run_14.xls", "SN1-21_Run_17.xls","SN1-23-Run_18.xls","SN1-25-Run_19.xls", "SN1-26-Run_20.xls")
resultFileNames <- c("Run_15.xls", "Run_16.xls","Run_17.xls","Run_18.xls", "Run_19.xls", "Run_20.xls")
for (i in 1:length(resultFileNames)){
  res <- readResultFile(resultFileNames[i], TRUE)
  namesOfColumns <- names(res$proportions)
  totalTable[namesOfColumns] <- res$proportions[namesOfColumns]
  namesOfColumns <- names(res$y)
  totalTableOfMeans[namesOfColumns] <- res$y
  totalTableOflogMeans[namesOfColumns] <- res$ylog
}
totalTable$Target <- y$Target
totalTable$sd <- apply(totalTable[, -1], 1, sd)
totalTable$mean <- apply(totalTable[, -1], 1, mean)
totalTableOfMeans$sd <- apply(totalTableOfMeans[, -1], 1, sd)
totalTableOflogMeans$sd <- apply(totalTableOflogMeans[, -1], 1, sd)
rownames(totalTable) <- totalTable$Target
rownames(totalTableOfMeans) <- totalTableOfMeans$Target
rownames(totalTableOflogMeans) <- totalTableOflogMeans$Target
```


(1) Get dangling ends
```{r}
prepared_data <- data.frame(Target = character(nrow(data)))
prepared_data$Target <- data$Amplicon_ID

prepared_data$Ten <- data$Amplicon_Inner_Part10
prepared_data$Start_Dangling_End <- data$Start_Dangling_End
prepared_data$Stop_Dangling_End <- data$Stop_Dangling_End
prepared_data$Start_Dangling_End <- as.factor(prepared_data$Start_Dangling_End)
prepared_data$Stop_Dangling_End <- as.factor(prepared_data$Stop_Dangling_End)

```

(1) Get lengths of amplicon and primers
```{r}
prepared_data$Fwd_Primer_Len <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Fwd_Primer.), function(x) nchar(x)))
prepared_data$Rev_Primer_Len <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Rev_Primer.), function(x) nchar(x)))
prepared_data$Amplicon_Len <- as.numeric(lapply(as.character(data$Amplicon_Inner_Part), function(x) nchar(x)))
```

(1) Get 3'-ends of amplicon
```{r}
prepared_data$Fwd_END_3 <- as.character(lapply(data$Amplicon_Inner_Part, function(x) substr(x, nchar(x), nchar(x))))
prepared_data$Rev_END_3 <- as.character(lapply(data$Amplicon_Inner_Part, function(x) substr(x, 1, 1)))
prepared_data$Fwd_END_3 <- as.factor(prepared_data$Fwd_END_3)
prepared_data$Rev_END_3 <- as.factor(prepared_data$Rev_END_3)
```

(1) Get GC-content
```{r}
countCharOccurrences <- function(char, s) {s2 <- gsub(char,"",s);return (nchar(s) - nchar(s2))}
countGCcontent <- function(x){ gg <- countCharOccurrences('G',x); cc <-countCharOccurrences('C',x); aa <- countCharOccurrences('A',x); tt <- countCharOccurrences('T',x); (cc+gg)*100/(aa+tt+cc+gg) }
prepared_data$Fwd_Primer_GC_content <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Fwd_Primer.), function(x)  countGCcontent(x)))
prepared_data$Rev_Primer_GC_content <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Rev_Primer.), function(x)  countGCcontent(x)))
prepared_data$Amplicon_GC_content <- as.numeric(lapply(as.character(data$Amplicon_Inner_Part), function(x) countGCcontent(x)))
```

(2) Get Max Poly-X (maximum allowable length of a mononucleotide repeat)
```{r, echo=FALSE}
getMaxPolyX <- function(str){
  n <- 0
  nmax <- 0
  p <- 'N'
  for (c in 1:nchar(str)){
    if (p == substr(str, c, c)) {
      n <- n+1
    }else{
      if (n > nmax) {
        nmax <- n
      }
      n <- 0
    }
    p <- substr(str, c, c)
  }
  nmax
}
```
```{r}
prepared_data$Fwd_Primer_MaxPolyX <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Fwd_Primer.), function(x) getMaxPolyX(x)))
prepared_data$Rev_Primer_MaxPolyX <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Rev_Primer.), function(x) getMaxPolyX(x)))
prepared_data$Amplicon_MaxPolyX <- as.numeric(lapply(as.character(data$Amplicon_Inner_Part), function(x) getMaxPolyX(x)))
```

(2) Get amplicon sequence entropy
```{r, echo=FALSE}
getEntropy <- function(str){
  h <- 0
  for (c in 1:nchar(str)){
    p <- substr(str, c, c)
    if (p == "A"){
      h <- h - 0.295*log(0.295)
    }
    if (p == "C"){
      h <- h - 0.205*log(0.205)
    }
    if (p == "G"){
      h <- h - 0.205*log(0.205)
    }
    if (p == "T"){
      h <- h - 0.295*log(0.295)
    }
  }
  h/nchar(str)
}
```
```{r}
prepared_data$Fwd_Primer_Entropy <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Fwd_Primer.), function(x) getEntropy(x)))
prepared_data$Rev_Primer_Entropy <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Rev_Primer.), function(x) getEntropy(x)))
prepared_data$Amplicon_Entropy <- as.numeric(lapply(as.character(data$Amplicon_Inner_Part), function(x) getEntropy(x)))
print(head(prepared_data$Amplicon_Entropy))
```

(2) Get hash of primer and amplicon sequences
```{r, echo=FALSE, warning=FALSE}
getHash <- function(str){
  hash <- 0
  mod <- 63018038201
  st <- 1
  for (c in 1:nchar(str)){
    p <- substr(str, c, c)
    if (p == "A"){
      hash <- hash %% mod
    }
    if (p == "C"){
      hash <- (hash + st) %% mod
    }
    if (p == "G"){
      hash <- (hash + 2*st) %% mod
    }
    if (p == "T"){
      hash <- (hash + 3*st) %% mod
    }
    st <- (st*4)%% mod
  }
  hash
}
```
```{r, warning=FALSE}
prepared_data$Fwd_Primer_Hash <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Fwd_Primer.), function(x) getHash(x)))
prepared_data$Rev_Primer_Hash <- as.numeric(lapply(as.character(data$Ion_AmpliSeq_Rev_Primer.), function(x) getHash(x)))
prepared_data$Amplicon_Hash <- as.numeric(lapply(as.character(data$Amplicon_Inner_Part), function(x) getHash(x)))
```

Get intersection of given amplicons and amplicons from experiments
```{r}
rownames(prepared_data) <- prepared_data$Target
valid_names <- intersect(y$Target, data$Amplicon_ID)
prepared_data <- prepared_data[valid_names,]
y <- y[valid_names,]
ylog <- ylog[valid_names,]
totalTable <- totalTable[valid_names,]
totalTableOfMeans <- totalTableOfMeans[valid_names,]
totalTableOflogMeans <- totalTableOflogMeans[valid_names,]
```

Prepare data for ThermoBLAST
```{r}
rownames(data) <- data$Amplicon_ID
valid_amplicons <- data[valid_names,]
thermoblast_data <- data.frame(Sequence = character(2*nrow(valid_amplicons)))
thermoblast_data$Sequence <- c(as.character(valid_amplicons$Ion_AmpliSeq_Fwd_Primer.), as.character(valid_amplicons$Ion_AmpliSeq_Rev_Primer.))
thermoblast_data$Concentration <- rep(0.0000001, nrow(thermoblast_data))
thermoblast_data$StrandType <- rep("DNA", nrow(thermoblast_data))
thermoblast_data$Function <- rep("PRIMER", nrow(thermoblast_data))
thermoblast_data$Name <- c(paste(valid_amplicons$Amplicon_ID, "_Fwd", sep=""), paste(valid_amplicons$Amplicon_ID, "_Rev", sep=""))
write.csv(thermoblast_data, "../data/panel2/ForThermoBLAST.csv", row.names=FALSE)
```

Save features to tsv file
```{r}
prepared_data$res <- y[,2]
prepared_data$logres <- ylog[,2]
prepared_data$sd <- totalTable$sd
prepared_data$mean <- totalTable$mean
prepared_data$MeansSD <- totalTableOfMeans$sd
write.table(prepared_data, "../data/panel2/features.tsv", col.names = TRUE, sep = "\t", row.names = FALSE, quote = FALSE)
```

